core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,776 @@
+		.sp_ns_comm_buf_base = ns_comm_buf_addr,
+		.sp_shared_buf_base = sec_buf_addr,
+		.sp_image_size = stmm_image_size,
+		.sp_pcpu_stack_size = stmm_stack_size,
+		.sp_heap_size = stmm_heap_size,
+		.sp_ns_comm_buf_size = stmm_ns_comm_buf_size,
+		.sp_shared_buf_size = stmm_sec_buf_size,
+		.num_sp_mem_regions = 6,
+		.num_cpus = 1,
+		.mp_info = mp_info,
+	};
+	mp_info->mpidr = read_mpidr_el1();
+	mp_info->linear_id = 0;
+	mp_info->flags = MP_INFO_FLAG_PRIMARY_CPU;
+	spc->ns_comm_buf_addr = ns_comm_buf_addr;
+	spc->ns_comm_buf_size = stmm_ns_comm_buf_size;
+
+	init_stmm_regs(spc, sec_buf_addr,
+		       (vaddr_t)(mp_info + 1) - sec_buf_addr,
+		       stack_addr + stmm_stack_size, image_addr);
+
+	return sec_part_enter_user_mode(spc);
+}
+
+TEE_Result sec_part_init_session(const TEE_UUID *uuid,
+				 struct tee_ta_session *sess)
+{
+	struct sec_part_ctx *spc = NULL;
+	TEE_Result res = TEE_SUCCESS;
+
+	if (memcmp(uuid, &stmm_uuid, sizeof(*uuid)))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	spc = sec_part_alloc_ctx(uuid);
+	if (!spc)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	spc->is_initializing = true;
+
+	sess->ctx = &spc->uctx.ctx;
+	tee_ta_push_current_session(sess);
+	res = load_stmm(spc);
+	tee_ta_pop_current_session();
+	tee_mmu_set_ctx(NULL);
+	if (res)
+		goto err;
+
+	spc->is_initializing = false;
+	TAILQ_INSERT_TAIL(&tee_ctxes, &spc->uctx.ctx, link);
+	return TEE_SUCCESS;
+
+err:
+	sess->ctx = NULL;
+	spc->uctx.ctx.ops->destroy(&spc->uctx.ctx);
+
+	return res;
+}
+
+static TEE_Result stmm_map_ns_buf(struct sec_part_ctx *spc,
+				  struct tee_ta_param *param,
+				  vaddr_t *ns_buf_base)
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+					  TEE_PARAM_TYPE_VALUE_OUTPUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+	void *param_va[TEE_NUM_PARAMS] = { NULL };
+	TEE_Result res = TEE_SUCCESS;
+
+	if (exp_pt != param->types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = tee_mmu_map_param(&spc->uctx, param, param_va);
+	if (!res)
+		*ns_buf_base = (vaddr_t)param_va[0];
+	return res;
+}
+
+static TEE_Result stmm_enter_open_session(struct tee_ta_session *s,
+					  struct tee_ta_param *param __unused,
+					  TEE_ErrorOrigin *eo)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(s->ctx);
+
+	if (spc->is_initializing) {
+		/* stmm is initialized in sec_part_init_session() */
+		*eo = TEE_ORIGIN_TEE;
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stmm_enter_invoke_cmd(struct tee_ta_session *s,
+					uint32_t cmd,
+					struct tee_ta_param *param,
+					TEE_ErrorOrigin *eo __unused)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(s->ctx);
+	TEE_Result res = TEE_SUCCESS;
+	vaddr_t ns_buf_base = 0;
+	unsigned int ns_buf_size;
+
+	if (cmd != PTA_STMM_COMMUNICATE)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
+	ns_buf_size = param->u[0].mem.size;
+	if (ns_buf_size > spc->ns_comm_buf_size)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = stmm_map_ns_buf(spc, param, &ns_buf_base);
```

Comment:
Why are we mapping the parameters in StMM's memory space when later doing a `memcpy()` into another buffer?

Reply:
As we discussed over google chat, this is the ns world buffer that comes from U-boot, which isn't always mapped. 
What you proposed using mobj_inc_map() works and it's better, since it adheres to existing APIs.
So I'll use these calls to map the buffer